src/ncp-spinel/SpinelNCPInstance.cpp

```diff
@@ -2556,93 +2539,183 @@ SpinelNCPInstance::handle_ncp_spinel_callback(unsigned int command, const uint8_
+	static const uint8_t rloc_bytes[] = {0x00,0x00,0x00,0xFF,0xFE,0x00};
+	bool should_filter = NCPInstanceBase::should_filter_address(addr, prefix_len);
 
-		factory.add_command(
-			SpinelPackData(
-				SPINEL_FRAME_PACK_CMD_PROP_VALUE_INSERT(
-					SPINEL_DATATYPE_IPv6ADDR_S   // Address
-					SPINEL_DATATYPE_UINT8_S      // Prefix Length
-					SPINEL_DATATYPE_BOOL_S       // Stable?
-					SPINEL_DATATYPE_UINT8_S      // Flags
-				),
-				SPINEL_PROP_THREAD_ON_MESH_NETS,
-				&addr,
-				prefix_len,
-				true,
-				flags
-			)
-		);
+	// Filter RLOC link-local or mesh-local addresses
 
-		start_new_task(factory.finish());
+	if (0 == memcmp(rloc_bytes, addr.s6_addr + 8, sizeof(rloc_bytes))) {
+		if (IN6_IS_ADDR_LINKLOCAL(&addr)) {
+			should_filter = true;
+		}
+
+		if (buffer_is_nonzero(mNCPV6Prefix, sizeof(mNCPV6Prefix))
+		    && (0 == memcmp(mNCPV6Prefix, &addr, sizeof(mNCPV6Prefix)))
+		) {
+			should_filter = true;
+		}
 	}
+
+	return should_filter;
 }
 
 void
-SpinelNCPInstance::address_was_removed(const struct in6_addr& addr, int prefix_len)
+SpinelNCPInstance::add_unicast_address_on_ncp(const struct in6_addr &addr, uint8_t prefix_len, CallbackWithStatus cb)
 {
-	if (mPrimaryInterface->is_online() && is_address_known(addr)) {
-		SpinelNCPTaskSendCommand::Factory factory(this);
-
-		factory.set_lock_property(SPINEL_PROP_THREAD_ALLOW_LOCAL_NET_DATA_CHANGE);
-
-		factory.add_command(
-			SpinelPackData(
-				SPINEL_FRAME_PACK_CMD_PROP_VALUE_REMOVE(
-					SPINEL_DATATYPE_IPv6ADDR_S   // Address
-					SPINEL_DATATYPE_UINT8_S      // Prefix
-				),
-				SPINEL_PROP_IPV6_ADDRESS_TABLE,
-				&addr,
-				prefix_len
-			)
-		);
+	SpinelNCPTaskSendCommand::Factory factory(this);
 
-		start_new_task(factory.finish());
-	}
+	syslog(LOG_NOTICE, "Adding address \"%s/%d\" to NCP", in6_addr_to_string(addr).c_str(), prefix_len);
 
-	NCPInstanceBase::address_was_removed(addr, prefix_len);
+	factory.set_lock_property(SPINEL_PROP_THREAD_ALLOW_LOCAL_NET_DATA_CHANGE);
+	factory.set_callback(cb);
+
+	factory.add_command(
+		SpinelPackData(
+			SPINEL_FRAME_PACK_CMD_PROP_VALUE_INSERT(
+				SPINEL_DATATYPE_IPv6ADDR_S   // Address
+				SPINEL_DATATYPE_UINT8_S      // Prefix Length
+				SPINEL_DATATYPE_UINT32_S     // Valid Lifetime
+				SPINEL_DATATYPE_UINT32_S     // Preferred Lifetime
+			),
+			SPINEL_PROP_IPV6_ADDRESS_TABLE,
+			&addr,
+			prefix_len,
+			UINT32_MAX,
+			UINT32_MAX
+		)
+	);
+
+	start_new_task(factory.finish());
 }
 
 void
-SpinelNCPInstance::check_operation_status(std::string operation, int status)
+SpinelNCPInstance::remove_unicast_address_on_ncp(const struct in6_addr& addr, uint8_t prefix_len, CallbackWithStatus cb)
 {
-	if (status == kWPANTUNDStatus_Timeout)
-	{
-		syslog(LOG_ERR, "Timed out while performing \"%s\" - Resetting NCP.", operation.c_str());
-		ncp_is_misbehaving();
-	}
+	SpinelNCPTaskSendCommand::Factory factory(this);
+
+	syslog(LOG_NOTICE, "Removing address \"%s/%d\" from NCP", in6_addr_to_string(addr).c_str(), prefix_len);
```

Comment:
I agree... Will change them to `LOG_INFO`. I think they were originally `LOG_NOTICE` that's why I kept them...